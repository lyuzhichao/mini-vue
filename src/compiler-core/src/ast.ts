export const enum NodeTypes {
    INTERPOLATION,
    SIMPLE_EXPRESSION,
    ELEMENT,
    TEXT,
    ROOT,
    COMPOND_EXPRESSION
}

export function createVNodeCall(context,vnodeTag,vnodeProps,vnodeChildren){
    return {
        type:NodeTypes.ELEMENT,
        tag:vnodeTag,
        props:vnodeProps,
        children:vnodeChildren
    }
}