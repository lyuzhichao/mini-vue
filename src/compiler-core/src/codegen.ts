import {NodeTypes} from "./ast";
import {CREATE_ELEMENT_VNODE, helperMapName, TO_DISPLAY_STRING} from "./runtimeHelper";
import {isString} from "../../shared";

export function generate(ast) {
    console.log('ast',ast)
    const context=createCodegenContext()
    const {push}=context

    getFunctionPreamble(ast,context)

    const functionName = 'render'
    const args = ['_ctx', '_cache','$props','$setup','$data','$options']
    const signature = args.join(', ')

    // const node=ast.codegenNode
    push(`function ${functionName}(${signature}){`)
    genNode(ast.codegenNode,context)
    push('}')
    console.log(context.code)
    return {
        code:context.code
    }


    // return {
    //     code: `return function render(_ctx,_cache,$props,$setup,$data,$options){return "hi"}`
    // }
}
function getFunctionPreamble(ast,context){
    const {push}=context
    const vueBing='Vue'
    const aliasHelper=(s)=>`${helperMapName[s]}: _${helperMapName[s]}`
    if (ast.helpers.length>0){
        push(`const { ${ast.helpers.map(aliasHelper).join(', ')} } = ${vueBing}`)
    }
    push('\n')
    push('return ')
}

function genNode(node,context){

    switch (node.type){
        case NodeTypes.TEXT:
            genText(node,context)
            break
        case NodeTypes.INTERPOLATION:
            genInterpolation(node,context)
            break
        case NodeTypes.SIMPLE_EXPRESSION:
            getExpression(node,context)
            break
        case NodeTypes.ELEMENT:
            genElement(node,context)
            break
        case NodeTypes.COMPOND_EXPRESSION:
            getCompoundExpression(node,context)
            break
        default:
            break
    }
}

function getCompoundExpression(node,context){
    const {push}=context
    const children=node.children
    for (let i=0;i<children.length;i++){
        const child =children[i]
        if (isString(child)){
            push(child)
        } else {
            genNode(child,context)
        }

    }
}

function genElement(node,context){
    const {push,helper}=context
    const {tag,children,props}=node
    push(`${helper(CREATE_ELEMENT_VNODE)}(`)

    genNodeList(genNullLable([tag,props,children]),context)

    push(")")
}

function genNodeList(nodes,context){
    const {push}=context
    for (let i=0;i<nodes.length;i++){
        const node=nodes[i]
        if (isString((node))){
            push(node)
        } else {
            genNode(node,context)
        }
        if (i<nodes.length-1){
            push(',')
        }
    }
}

function genNullLable(args){
    return args.map((arg)=>arg || "null")
}

function genText(node,context){
    const {push}=context
    push(`"${node.content}"`)
}

function genInterpolation(node,context){
    const {push,helper} = context
    push(`${helper(TO_DISPLAY_STRING)}(`)
    genNode(node.content,context)
    push(')')
}

function getExpression(node,context){
    const {push} = context
    push(`${node.content}`)
}

function createCodegenContext(){
    const context={
        code:'',
        push(source){
            context.code+=source
        },
        helper(key){
            return `_${helperMapName[key]}`
        }
    }
    return context
}