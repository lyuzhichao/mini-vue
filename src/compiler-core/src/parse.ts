import {NodeTypes} from "./ast";

const enum TagType {
    Start,
    End
}

export function baseParse(content: string) {

    const context = createParseContext(content)

    return createRoot(parseChildren(context))
}

function parseChildren(context) {

    const nodes:any=[]
    let node
    const s=context.source
    if (s.startsWith("{{")){
        node=parseInterpolation(context)
    } else if (s[0]==='<'){
        if (/[a-z]/i.test(s[1])){
            node=parseElement(context)
            console.log('is element')
        }
    }

    if (!node){
        node=parseText(context)
    }

    nodes.push(node)
    return nodes
}

function parseTextData(context,length){
    const content=context.source.slice(0,length)
    advanceBy(context,length)
    return content
}

function parseText(context){

    // const content=context.source.slice(0,context.source.length)
    // advanceBy(context,content.length)
    const content=parseTextData(context,context.source.length)
    console.log('text: ',context.source)
    return {
        type:NodeTypes.TEXT,
        tag:content
    }
}

function parseTag(context,tagType){
    const match:any=/^<\/?([a-z]*)/i.exec(context.source)
    console.log('match: ',match)
    const tag=match[1]
    advanceBy(context,match[0].length)
    advanceBy(context,1)
    if (tagType===TagType.End) return
    console.log('advabce element source',context.source)
    return {
        type:NodeTypes.ELEMENT,
        tag:tag
    }
}

function parseElement(context){
    const element =  parseTag(context,TagType.Start)
    parseTag(context,TagType.End)
    console.log('-----------------',context.source)
    return element
}

function parseInterpolation(context) {
    const openDelimiter="{{"
    const closeDelimiter="}}"
    const closeIndex=context.source.indexOf(closeDelimiter,openDelimiter.length)

    advanceBy(context,openDelimiter.length)

    console.log('context source',context.source)

    const rawContentLength=closeIndex-openDelimiter.length
    const rawContent=parseTextData(context,rawContentLength)
    console.log('rawContent: ',rawContent)
    const content=rawContent.trim()
    console.log('content: ',content)
    advanceBy(context,rawContentLength+closeDelimiter.length)
    console.log('remove source: ',context.source)

    return {
        type: NodeTypes.INTERPOLATION,
        content: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: content
        }
    }
}
function advanceBy(context,length){
    context.source=context.source.slice(length)
}


function createParseContext(content: string) {
    return {
        source: content
    }
}

function createRoot(children) {
    return {
        children
    }
}