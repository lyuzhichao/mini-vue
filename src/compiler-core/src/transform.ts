import {NodeTypes} from "./ast";
import {TO_DISPLAY_STRING} from "./runtimeHelper";

export function transform(root,options={}){
    console.log('original root: ',root)
    const context=createTransformContext(root,options)
    //深度优先搜索
    tranverseNode(root,context)
    createRootCodegen(root)
    root.helpers=[...context.helpers.keys()]
    console.log('root: ',root)
}

function createRootCodegen(root){
    if (root.children){
        root.codegenNode=root.children[0]
    }
}

function tranverseNode(node,context){

    const nodeTransforms=context.nodeTransforms
    for (let i=0;i<nodeTransforms.length;i++){
        const transform=nodeTransforms[i]
        transform(node)
    }

    switch (node.type){
        case NodeTypes.INTERPOLATION:
            context.helper(TO_DISPLAY_STRING)
            break
        case NodeTypes.ROOT:
        case NodeTypes.ELEMENT:
            tranverseChildren(node,context)
            break
        default:
            break
    }

    // tranverseChildren(node,context)

}

function tranverseChildren(parent,context){
    const children=parent.children
    if(children){
        for (let i=0;i<children.length;i++){
            const node=children[i]
            tranverseNode(node,context)
        }
    }
}

function createTransformContext(root,options){
    const context={
        root,
        nodeTransforms:options.nodeTransforms || [],
        helpers:new Map(),
        helper(key){
            context.helpers.set(key,1)
        }
    }
    return context
}