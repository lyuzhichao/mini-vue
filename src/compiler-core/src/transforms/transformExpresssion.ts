import {NodeTypes} from "../ast";

export function transformExpresssion(node){
    if (node.type===NodeTypes.INTERPOLATION){
        node.content=processExpression(node.content)
    }
}

function processExpression(node){
    node.content=`ctx_.${node.content}`
    return node
}
