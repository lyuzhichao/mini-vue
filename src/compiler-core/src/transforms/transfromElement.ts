import {createVNodeCall, NodeTypes} from "../ast";
import {CREATE_ELEMENT_VNODE} from "../runtimeHelper";

export function transformElement(node,context){
    if (node.type===NodeTypes.ELEMENT){
        return ()=>{
            context.helper(CREATE_ELEMENT_VNODE)

            const vnodeTag=`"${node.tag}"`

            let vNodeProps

            const children = node.children
            let vnodeChildren=children[0]

            // const vnodeElement = {
            //     type:NodeTypes.ELEMENT,
            //     tag:vnodeTag,
            //     props:vNodeProps,
            //     children:vnodeChildren
            // }

            node.codegenNode=createVNodeCall(
                context,
                vnodeTag,
                vNodeProps,
                vnodeChildren
            )
        }
    }
}