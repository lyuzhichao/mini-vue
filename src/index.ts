//mini-vue output
// export * from './runtime-core'
export * from './runtime-dom'
export * from './reactivity'

import * as runtimedom from './runtime-dom'
import {baseCompile} from "./compiler-core/src";
import {registerRuntimeCompiler} from "./runtime-dom";

function compileToFunction(template) {
    const {code} = baseCompile(template)

    const render = new Function('Vue',code)(runtimedom)

    return render
    // function renderFunction(Vue) {
    //     const {toDisplayString: _toDisplayString, createElementVNode: _createElementVNode, openBlock: _openBlock} = Vue
    //     return function render(_ctx, _cache, $props, $setup, $data, $options){
    //         _createElementVNode("div",null,"hi, "+_toDisplayString(ctx_.message))
    //     }
    //
    // }
}
registerRuntimeCompiler(compileToFunction)