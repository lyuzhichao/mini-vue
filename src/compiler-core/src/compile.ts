import {baseParse} from "./parse";
import {transform} from "./transform";
import {generate} from "./codegen";
import {transformExpresssion} from "./transforms/transformExpresssion";
import {transformElement} from "./transforms/transfromElement";
import {transformText} from "./transforms/transformText";

export function baseCompile(template){
    const ast=baseParse(template)
    transform(ast,{
        nodeTransforms:[transformExpresssion,transformElement,transformText]
    })
    return generate(ast)
}

