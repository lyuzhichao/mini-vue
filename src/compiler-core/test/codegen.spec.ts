import {baseParse} from "../src/parse";
import {generate} from "../src/codegen";
import {transform} from "../src/transform";
import {transformExpresssion} from "../src/transforms/transformExpresssion";
import {transformElement} from "../src/transforms/transfromElement";
import {transformText} from "../src/transforms/transformText";

describe('codegen',()=>{
    it('string',()=>{
        const ast=baseParse('hi')
        transform(ast)
        const {code}=generate(ast)
        //snapShot
        expect(code).toMatchSnapshot()
    });
    it('interpolation',()=>{
        const ast=baseParse('{{message}}')
        transform(ast,{
            nodeTransforms:[transformExpresssion]
        })
        const {code}=generate(ast)
        //snapShot
        expect(code).toMatchSnapshot()
    });
    it('element',()=>{
        const ast=baseParse("<div>hi, {{message}}</div>")
        transform(ast,{
            nodeTransforms:[transformExpresssion,transformElement,transformText]
        })
        const {code}=generate(ast)
        //snapShot
        expect(code).toMatchSnapshot()
    })
})