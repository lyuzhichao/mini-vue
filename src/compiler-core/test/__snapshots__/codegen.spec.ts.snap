// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
"const { toDisplayString: _toDisplayString, createElementVNode: _createElementVNode } = Vue
return function render(_ctx, _cache, $props, $setup, $data, $options){_createElementVNode("div",null,"hi, "+_toDisplayString(ctx_.message))}"
`;

exports[`codegen interpolation 1`] = `
"const { toDisplayString: _toDisplayString } = Vue
return function render(_ctx, _cache, $props, $setup, $data, $options){_toDisplayString(ctx_.message)}"
`;

exports[`codegen string 1`] = `
"
return function render(_ctx, _cache, $props, $setup, $data, $options){"hi"}"
`;
